### PR TITLE
 tree-wide: make rust jemalloc-sys use nixpkgs jemalloc build

### PR DESCRIPTION
**Pull request by yu-re-ka**
###### Description of changes

Addresses #202863

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by yu-re-ka** [comment]
@ofborg build pkgsStatic.fd

**Comment by yu-re-ka** [comment]
@ofborg build fd

**Comment by yu-re-ka** [comment]
More difficult than I thought because some packages (where the `unprefixed_jemalloc_on_supported_platforms` feature is enabled) are not allowed to have the `_rjem_` prefix depending on the platform.

**Comment by tjni** [comment]
What's your current thinking on how to address this? It seems hard to do automatically without better ways to override.

**Comment by yu-re-ka** [comment]
We can create a prefixed (as is currently in this PR) and only-prefixed-on-some-platforms version.
However we need to exactly match the list [here](https://github.com/tikv/jemallocator/blob/0fef73074b57e8b1e6abb3d6beedfb823fce01bc/jemalloc-sys/src/env.rs#L17) and hope that it doesn't change too much across versions of the (tikv-)jemallocator crates.

Then choosing the correct version of the hook should be a matter of trial-and-error.

**Comment by tjni** [comment]
This does sound like the best option today.

> However for the only-prefixed-on-some-platforms version we need to exactly match the list

Perhaps this can be considered an issue that upstreams would be amenable to fixing, so then our second variant is just the system `jemalloc`.

> Then another challenge is finding all rust packages that build with jemallocator

I think the only way to find these right now is if someone like you encounters a breakage.

File: pkgs/development/libraries/jemalloc/rust.nix [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  ];
+
+  postFixup = ''
+    touch $out/rust-jemalloc
+  '';
+
+  setupHook = writeText "setup-hook.sh" ''
+    addRustJemalloc () {
+      if test -f "$1/rust-jemalloc"; then
+        export JEMALLOC_OVERRIDE="$1/lib/libjemalloc${stdenv.hostPlatform.extensions.library}"
```
Comment (montchr) [inline pkgs/development/libraries/jemalloc/rust.nix:16]: I think I'm probably missing something, but would `export JEMALLOC_SYS_WITH_LG_PAGE=16` not work here?

Comment (yu-re-ka) [reply pkgs/development/libraries/jemalloc/rust.nix:16]: If you look at the options passed `pkgs/development/libraries/jemalloc/default.nix`, they are applying more fixes specifically for arm 32-bit, x86 darwin, and enabling up to 64K pagesize.
We could pass all of these options, but  then it gets more complicated to keep them in sync, compared to just one JEMALLOC_OVERRIDE.

File: pkgs/development/libraries/jemalloc/rust.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
```
Comment (alyssais) [inline pkgs/development/libraries/jemalloc/rust.nix:28]: I think we can maybe do

```suggestion
      export JEMALLOC_OVERRIDE="@out@/lib/libjemalloc${stdenv.hostPlatform.extensions.library}"
```

and not need the `if` or the `postFixup`?

**Review by alyssais — COMMENTED** [review]
The unprefixed variant doesn't seem to be used anywhere — is that still waiting for testing each package to determine which one it needs?

File: pkgs/development/libraries/jemalloc/rust.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+in jemalloc.overrideAttrs (oldAttrs: {
+  configureFlags = oldAttrs.configureFlags ++ lib.optionals (!unprefixed') [
+    "--with-jemalloc-prefix=_rjem_"
+  ] ++ [
+    "--with-private-namespace=_rjem_"
+  ];
+
+  setupHook = writeText "setup-hook.sh" ''
+    export JEMALLOC_OVERRIDE="@out@/lib/libjemalloc${stdenv.hostPlatform.extensions.library}"
+  '';
```
Comment (alyssais) [inline pkgs/development/libraries/jemalloc/rust.nix:23]: I wonder, what if we put this in the normal jemalloc package?  It seems a shame to rebuild (for unprefixed) _just_ to add a setup hook.

Comment (yu-re-ka) [reply pkgs/development/libraries/jemalloc/rust.nix:23]: The unprefixed variant still needs a `"--with-private-namespace=_rjem_"`

Comment (alyssais) [reply pkgs/development/libraries/jemalloc/rust.nix:23]: Ahh I misread.  Maybe you could swap the order --with-jemalloc-prefix and --with-private-namespace, so it's clearer that only --with-jemalloc-prefix is conditionally applied?

Comment (yu-re-ka) [reply pkgs/development/libraries/jemalloc/rust.nix:23]: done

**Review by alyssais — APPROVED** [review]

**Comment by hmenke** [comment]
This broke all Rust static builds. For example `ruff`
```console
$ git checkout 799fff889b1e
$ nix-build -A pkgsStatic.ruff # error
$ git checkout HEAD^
$ nix-build -A pkgsStatic.ruff # succeeds
```

**Comment by alyssais** [comment]
Right, yes.  We have two options, then.  Make every package only conditionally depend on rust-jemalloc-sys when !stdenv.hostPlatform.isStatic, or set rust-jemalloc-sys to null in that case.  @yu-re-ka, any thoughts?

**Pull request merged**